main.go

```diff
@@ -181,37 +181,40 @@ func runDaemon() error {
 
 	// initialize signal and handle method.
 	var (
-		waitExit = make(chan struct{})
-		signals  = make(chan os.Signal, 1)
+		errCh    = make(chan error, 1)
+		signalCh = make(chan os.Signal, 1)
 	)
-	signal.Notify(signals, syscall.SIGINT, syscall.SIGQUIT, syscall.SIGTERM, syscall.SIGHUP)
-	go func() {
-		sig := <-signals
-		logrus.Warnf("received signal: %s", sig)
-
-		for _, handle := range sigHandles {
-			if err := handle(); err != nil {
-				logrus.Errorf("failed to handle signal: %v", err)
-			}
-		}
-
-		close(waitExit)
-		os.Exit(1)
-	}()
 
 	// new daemon instance, this is core.
 	d := daemon.NewDaemon(cfg)
 	if d == nil {
 		return fmt.Errorf("failed to new daemon")
 	}
 
+	signal.Notify(signalCh, syscall.SIGINT, syscall.SIGQUIT, syscall.SIGTERM, syscall.SIGHUP)
 	sigHandles = append(sigHandles, d.ShutdownPlugin, d.Shutdown)
 
-	err := d.Run()
+	go func() {
+		// FIXME: I think the Run() should always return error.
+		errCh <- d.Run()
+	}()
 
-	<-waitExit
+	select {
+	case sig := <-signalCh:
+		logrus.Warnf("received signal: %s", sig)
 
-	return err
+		for _, handle := range sigHandles {
+			if err := handle(); err != nil {
+				logrus.Errorf("failed to handle signal: %v", err)
+			}
+		}
+
+		os.Exit(1)
+	case err := <-errCh:
+		// FIXME: should we do the cleanup like signal handle?
```

Comment:
I think we also should do cleanup when daemon exited.  
I do not know if this case will happen:  pouchd panic and return error, then we should make the daemon exit gracefully. WDYT ?

Reply:
Agree. But for now, we should hold the scope like @rudyfly said. Refactor the code in the following PR. WDYT?

Reply:
Confirm with @HusterWan , let us add TODO here.